File: cpp/tests/test_binary_heap.cpp
#include <gtest/gtest.h>

#include <algorithm>
#include <stdexcept>
#include <vector>

#include "binary_heap.h"

namespace {

template <typename T>
bool is_min_heap(const std::vector<T>& data) {
    for (std::size_t index = 0; index < data.size(); ++index) {
        std::size_t left_child_index = 2 * index + 1;
        std::size_t right_child_index = 2 * index + 2;

        if (left_child_index < data.size() && data[index] > data[left_child_index]) {
            return false;
        }

        if (right_child_index < data.size() && data[index] > data[right_child_index]) {
            return false;
        }
    }

    return true;
}

std::vector<int> pop_all(BinaryHeap<int>& heap) {
    std::vector<int> result;

    while (!heap.empty()) {
        result.push_back(heap.pop());
    }

    return result;
}

} // namespace

TEST(BinaryHeapTest, PushAndPeekBasic) {
    BinaryHeap<int> heap;

    heap.push(5);
    heap.push(2);
    heap.push(8);

    EXPECT_EQ(heap.peek(), 2);
    EXPECT_EQ(heap.size(), 3);
}

TEST(BinaryHeapTest, PopReturnsMinElement) {
    BinaryHeap<int> heap;

    heap.push(5);
    heap.push(2);
    heap.push(8);

    EXPECT_EQ(heap.pop(), 2);
    EXPECT_EQ(heap.size(), 2);
    EXPECT_EQ(heap.peek(), 5);
}

TEST(BinaryHeapTest, PopAllReturnsSortedOrder) {
    BinaryHeap<int> heap;
    std::vector<int> values = {5, 1, 8, 3, 2, 10};

    for (int value : values) {
        heap.push(value);
    }

    std::vector<int> result = pop_all(heap);
    std::sort(values.begin(), values.end());

    EXPECT_EQ(result, values);
    EXPECT_EQ(heap.size(), 0);
}

TEST(BinaryHeapTest, PushReverseOrder) {
    BinaryHeap<int> heap;
    std::vector<int> values = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};

    for (int value : values) {
        heap.push(value);
    }

    std::vector<int> result = pop_all(heap);
    std::sort(values.begin(), values.end());

    EXPECT_EQ(result, values);
}

TEST(BinaryHeapTest, PushWithDuplicates) {
    BinaryHeap<int> heap;
    std::vector<int> values = {5, 1, 3, 1, 5, 2};

    for (int value : values) {
        heap.push(value);
    }

    std::vector<int> result = pop_all(heap);
    std::sort(values.begin(), values.end());

    EXPECT_EQ(result, values);
}

TEST(BinaryHeapTest, HeapifyBuildsValidHeap) {
    std::vector<int> values = {5, 1, 8, 3, 2, 10};

    BinaryHeap<int> heap(values);

    EXPECT_TRUE(is_min_heap(heap.to_list()));
    EXPECT_EQ(heap.size(), values.size());
}

TEST(BinaryHeapTest, HeapifyPopAllReturnsSortedOrder) {
    std::vector<int> values = {5, 1, 8, 3, 2, 10};

    BinaryHeap<int> heap(values);

    std::vector<int> result = pop_all(heap);
    std::sort(values.begin(), values.end());

    EXPECT_EQ(result, values);
}

TEST(BinaryHeapTest, HeapifyEmptyList) {
    BinaryHeap<int> heap(std::vector<int>{});

    EXPECT_EQ(heap.size(), 0);
    EXPECT_TRUE(heap.to_list().empty());
}

TEST(BinaryHeapTest, HeapifySingleElement) {
    BinaryHeap<int> heap(std::vector<int>{42});

    EXPECT_EQ(heap.size(), 1);
    EXPECT_EQ(heap.peek(), 42);
    EXPECT_EQ(heap.pop(), 42);
    EXPECT_EQ(heap.size(), 0);
}

TEST(BinaryHeapTest, PeekEmptyHeapThrows) {
    BinaryHeap<int> heap;

    EXPECT_THROW(heap.peek(), std::out_of_range);
}

TEST(BinaryHeapTest, PopEmptyHeapThrows) {
    BinaryHeap<int> heap;

    EXPECT_THROW(heap.pop(), std::out_of_range);
}

TEST(BinaryHeapTest, SizeTracksNumberOfElements) {
    BinaryHeap<int> heap;

    EXPECT_EQ(heap.size(), 0);

    heap.push(10);
    heap.push(20);
    heap.push(5);

    EXPECT_EQ(heap.size(), 3);

    heap.pop();

    EXPECT_EQ(heap.size(), 2);

    heap.pop();
    heap.pop();

    EXPECT_EQ(heap.size(), 0);
}

TEST(BinaryHeapTest, ContainsExistingValue) {
    BinaryHeap<int> heap;

    heap.push(10);
    heap.push(5);
    heap.push(20);

    EXPECT_TRUE(heap.contains(5));
    EXPECT_TRUE(heap.contains(10));
    EXPECT_TRUE(heap.contains(20));
}

TEST(BinaryHeapTest, ContainsMissingValue) {
    BinaryHeap<int> heap;

    heap.push(10);
    heap.push(5);

    EXPECT_FALSE(heap.contains(99));
}

TEST(BinaryHeapTest, EmptyHeapIsEmpty) {
    BinaryHeap<int> heap;

    EXPECT_TRUE(heap.empty());
}

TEST(BinaryHeapTest, NonEmptyHeapIsNotEmpty) {
    BinaryHeap<int> heap;

    heap.push(1);

    EXPECT_FALSE(heap.empty());
}

TEST(BinaryHeapTest, ToListReturnsCopy) {
    BinaryHeap<int> heap;

    heap.push(3);
    heap.push(1);
    heap.push(2);

    std::vector<int> data = heap.to_list();
    data.push_back(100);

    EXPECT_EQ(heap.size(), 3);
    EXPECT_FALSE(heap.contains(100));
}

TEST(BinaryHeapTest, CopyPreservesAllElements) {
    BinaryHeap<int> heap;
    std::vector<int> values = {5, 1, 8, 3, 2};

    for (int value : values) {
        heap.push(value);
    }

    BinaryHeap<int> heap_copy = heap.copy();

    std::vector<int> result = pop_all(heap_copy);
    std::sort(values.begin(), values.end());

    EXPECT_EQ(result, values);
    EXPECT_EQ(heap.size(), values.size());
}

TEST(BinaryHeapTest, CopyIsIndependentFromOriginal) {
    BinaryHeap<int> heap;

    heap.push(5);
    heap.push(1);
    heap.push(3);

    BinaryHeap<int> heap_copy = heap.copy();

    heap.pop();
    heap_copy.push(0);

    EXPECT_EQ(heap.peek(), 3);
    EXPECT_EQ(heap_copy.peek(), 0);
}

TEST(BinaryHeapTest, MixedPushPopOperations) {
    BinaryHeap<int> heap;

    heap.push(10);
    heap.push(4);
    heap.push(7);

    EXPECT_EQ(heap.pop(), 4);

    heap.push(1);
    heap.push(9);

    EXPECT_EQ(heap.pop(), 1);
    EXPECT_EQ(heap.pop(), 7);
    EXPECT_EQ(heap.pop(), 9);
    EXPECT_EQ(heap.pop(), 10);
    EXPECT_EQ(heap.size(), 0);
}

TEST(BinaryHeapTest, HeapPropertyAfterEachPush) {
    BinaryHeap<int> heap;
    std::vector<int> values = {9, 4, 7, 1, 3, 8, 2};

    for (int value : values) {
        heap.push(value);
        EXPECT_TRUE(is_min_heap(heap.to_list()));
    }
}

TEST(BinaryHeapTest, HeapPropertyAfterEachPop) {
    BinaryHeap<int> heap(std::vector<int>{9, 4, 7, 1, 3, 8, 2});

    while (!heap.empty()) {
        heap.pop();
        EXPECT_TRUE(is_min_heap(heap.to_list()));
    }
}

File: python/tests/test_binary_heap.py
import pytest

from binary_heap import BinaryHeap


def is_min_heap(data: list) -> bool:
    for index in range(len(data)):
        left_child_index = 2 * index + 1
        right_child_index = 2 * index + 2

        if left_child_index < len(data) and data[index] > data[left_child_index]:
            return False

        if right_child_index < len(data) and data[index] > data[right_child_index]:
            return False

    return True


def test_push_and_peek_basic():
    heap = BinaryHeap()

    heap.push(5)
    heap.push(2)
    heap.push(8)

    assert heap.peek() == 2
    assert len(heap) == 3


def test_pop_returns_min_element():
    heap = BinaryHeap()

    heap.push(5)
    heap.push(2)
    heap.push(8)

    assert heap.pop() == 2
    assert len(heap) == 2
    assert heap.peek() == 5


def test_pop_all_returns_sorted_order():
    heap = BinaryHeap()

    values = [5, 1, 8, 3, 2, 10]

    for value in values:
        heap.push(value)

    result = []

    while len(heap) > 0:
        result.append(heap.pop())

    assert result == sorted(values)
    assert len(heap) == 0


def test_push_reverse_order():
    heap = BinaryHeap()

    values = [10, 9, 8, 7, 6, 5, 4, 3, 2, 1]

    for value in values:
        heap.push(value)

    result = []

    while len(heap) > 0:
        result.append(heap.pop())

    assert result == sorted(values)


def test_push_with_duplicates():
    heap = BinaryHeap()

    values = [5, 1, 3, 1, 5, 2]

    for value in values:
        heap.push(value)

    result = []

    while len(heap) > 0:
        result.append(heap.pop())

    assert result == sorted(values)


def test_heapify_builds_valid_heap():
    values = [5, 1, 8, 3, 2, 10]

    heap = BinaryHeap(values)

    assert is_min_heap(heap.to_list())
    assert len(heap) == len(values)


def test_heapify_pop_all_returns_sorted_order():
    values = [5, 1, 8, 3, 2, 10]

    heap = BinaryHeap(values)

    result = []

    while len(heap) > 0:
        result.append(heap.pop())

    assert result == sorted(values)


def test_heapify_empty_list():
    heap = BinaryHeap([])

    assert len(heap) == 0
    assert heap.to_list() == []


def test_heapify_single_element():
    heap = BinaryHeap([42])

    assert len(heap) == 1
    assert heap.peek() == 42
    assert heap.pop() == 42
    assert len(heap) == 0


def test_peek_empty_heap_raises_indexerror():
    heap = BinaryHeap()

    with pytest.raises(IndexError):
        heap.peek()


def test_pop_empty_heap_raises_indexerror():
    heap = BinaryHeap()

    with pytest.raises(IndexError):
        heap.pop()


def test_len_tracks_number_of_elements():
    heap = BinaryHeap()

    assert len(heap) == 0

    heap.push(10)
    heap.push(20)
    heap.push(5)

    assert len(heap) == 3

    heap.pop()

    assert len(heap) == 2

    heap.pop()
    heap.pop()

    assert len(heap) == 0


def test_contains_existing_value():
    heap = BinaryHeap()

    heap.push(10)
    heap.push(5)
    heap.push(20)

    assert 5 in heap
    assert 10 in heap
    assert 20 in heap


def test_contains_missing_value():
    heap = BinaryHeap()

    heap.push(10)
    heap.push(5)

    assert 99 not in heap


def test_bool_empty_heap_is_false():
    heap = BinaryHeap()

    assert not heap


def test_bool_non_empty_heap_is_true():
    heap = BinaryHeap()

    heap.push(1)

    assert heap


def test_to_list_returns_copy():
    heap = BinaryHeap()

    heap.push(3)
    heap.push(1)
    heap.push(2)

    data = heap.to_list()
    data.append(100)

    assert len(heap) == 3
    assert 100 not in heap


def test_copy_preserves_all_elements():
    heap = BinaryHeap()

    values = [5, 1, 8, 3, 2]

    for value in values:
        heap.push(value)

    heap_copy = heap.copy()

    result = []

    while len(heap_copy) > 0:
        result.append(heap_copy.pop())

    assert result == sorted(values)
    assert len(heap) == len(values)


def test_copy_is_independent_from_original():
    heap = BinaryHeap()

    heap.push(5)
    heap.push(1)
    heap.push(3)

    heap_copy = heap.copy()

    heap.pop()
    heap_copy.push(0)

    assert heap.peek() == 3
    assert heap_copy.peek() == 0


def test_mixed_push_pop_operations():
    heap = BinaryHeap()

    heap.push(10)
    heap.push(4)
    heap.push(7)

    assert heap.pop() == 4

    heap.push(1)
    heap.push(9)

    assert heap.pop() == 1
    assert heap.pop() == 7
    assert heap.pop() == 9
    assert heap.pop() == 10
    assert len(heap) == 0


def test_heap_property_after_each_push():
    heap = BinaryHeap()

    values = [9, 4, 7, 1, 3, 8, 2]

    for value in values:
        heap.push(value)
        assert is_min_heap(heap.to_list())


def test_heap_property_after_each_pop():
    heap = BinaryHeap([9, 4, 7, 1, 3, 8, 2])

    while len(heap) > 0:
        heap.pop()
        assert is_min_heap(heap.to_list())